docs/userguide/joins.rst

```diff
@@ -0,0 +1,260 @@
+Join Apps
+=========
+
+Join apps allows an app to define a sub-workflow: the app can launch other apps
+and incorporate them into the main task graph. They can be specified using the
+`join_app` decorator.
+
+Usage
+-----
+
+A `join_app` looks quite like a python app, but should return a ``Future``,
+rather than a value. After the python code has run, the app invocation will not
+complete until that future has completed, and the return value of the `join_app`
+will be the return value (or exception) from the returned future.
+
+In the following example, a sub-workflow called ``process`` is defined. This
+workflow decides which app to run on its argument, type_one if the input
+is positive and type_two if the input is negative.
+
+.. code-block:: python
+
+  @python_app
+  def type_one(x):
+    # do some stuff
+    return x*2
+
+  @python_app
+  def type_two(x):
+    # do some more stuff
+    return (-x) * 2
+
+  @join_app
+  def process(x):
+    if x > 0:
+      return type_one(x)
+    else:
+      return type_two(x)
+
+  # process(10).result() == 20
+  # process(-3).result() = 6
+
+The join app ``process`` looks very much like a normal Python function
+which launches apps would. When invoked with literal numbers as above,
+it would behave as if the `join_app` decorator wasn't there: the literal
+number is inspected and a choice is made about which app to invoke;
+then the ``Future`` that is returned from ``process`` completes when
+processing is finished.
+
+The difference comes when the input ``x`` is to come from the output of
+an earlier app invocation:
+
+.. code-block:: python
+
+  @python_app
+  def later():
+    time.sleep(120)
+    return 10
+
+  # process(later()) = 20
+
+If ``process`` was a regular Python function, it would be able to inspect
+the result of ``later`` - that result wouldn't exist for another 120 seconds.
+
+A join app would defer its execution until ``later`` was complete, just like
+any other Parsl app that is passed a ``Future``.
```

Comment:
again, at this point it's not clear why something different than a python_app is needed